File: leetcode/Algorithms/397.IntegerReplacement/Solution.c
int integerReplacement(int n) {
    int result = 0;

    while (n != 1) {
        if ((n & 1) == 0) {
            n >>= 1;
            result++;
        } else if ((n == 3) || ((n >> 1) & 1) == 0) {
            n = (n - 1) / 2;
            result += 2;
        } else {
            n = (n - 1) / 2 + 1;
            result += 2;
        }
    }

    return result;
}

File: leetcode/Algorithms/397.IntegerReplacement/Solution.java
public class Solution {
    public int integerReplacement(int n) {
        if (n == 1) return 0;
        if (n % 2 == 0) return integerReplacement(n / 2) + 1;
        return Math.min(integerReplacement(n - 1) + 1, integerReplacement((n - 1) / 2 + 1) + 2);
    }
}
